astropy/modeling/tests/test_quantities_parameters.py

```diff
@@ -0,0 +1,476 @@
+    assert exc.value.args[0] == ("Cannot set parameter units to Jy since it is "
+                                 "not equivalent with the original units of Hz")
+
+
+def test_parameter_set_value():
+    """
+    Test that changing the value on a parameter works as expected.
+    """
+
+    g = Gaussian1D(1 * u.Jy, 1 * u.m, 0.1 * u.m)
+
+    # To set a parameter to a quantity, we simply do
+    g.amplitude = 2 * u.Jy
+
+    # If we try setting the value, we need to pass a non-quantity value
+    # TODO: determine whether this is the desired behavior?
+    g.amplitude.value = 4
+    assert_quantity_allclose(g.amplitude, 4 * u.Jy)
+    assert g.amplitude.value == 4
+    assert g.amplitude.unit is u.Jy
+
+    # If we try setting it to a Quantity, we raise an error
+    with pytest.raises(TypeError) as exc:
+        g.amplitude.value = 3 * u.Jy
+    assert exc.value.args[0] == ("The .value property on parameters should be set to "
+                                 "unitless values, not Quantity objects. To set a "
+                                 "parameter to a quantity simply set the parameter "
+                                 "directly without using .value")
+
+
+def test_parameter_quantity_property():
+    """
+    Test that the quantity property of Parameters behaves as expected
+    """
+
+    # Since parameters have a .value and .unit parameter that return just the
+    # value and unit respectively, we also have a .quantity parameter that
+    # returns a Quantity instance.
+
+    g = Gaussian1D(1 * u.Jy, 1 * u.m, 0.1 * u.m)
+
+    assert_quantity_allclose(g.amplitude.quantity, 1 * u.Jy)
+
+    # Setting a parameter to a quantity changes the value and the default unit
+    g.amplitude.quantity = 5 * u.mJy
+    assert g.amplitude.value == 5
+    assert g.amplitude.unit is u.mJy
+
+    # But we shouldn't be able to set .quantity to a Quantity with
+    # non-equivalent units
+    with pytest.raises(UnitsError) as exc:
+        g.amplitude.quantity = 3 * u.m
+    assert exc.value.args[0] == ("Cannot set parameter units to m since it is not "
+                                 "equivalent with the original units of mJy")
+
+
+def test_parameter_default_units_match():
+
+    # If the unit and default quantity units are different, raise an error
+    with pytest.raises(ParameterDefinitionError) as exc:
+        class TestC(Fittable1DModel):
+            a = Parameter(default=1.0 * u.m, unit=u.Jy)
+    assert exc.value.args[0] == ("parameter default 1.0 m does not have units "
+                                 "equivalent to the required unit Jy")
+
+
+
+@pytest.mark.parametrize(('unit', 'default'), ((u.m, 1.0), (None, 1 * u.m)))
+def test_parameter_defaults(unit, default):
+    """
+    Test that default quantities are correctly taken into account
+    """
+
+    class TestModel(BaseTestModel):
+        a = Parameter(default=default, unit=unit)
+
+    # TODO: decide whether the default property should return a value or
+    # a quantity?
```

Comment:
If a unit is provided, I would say a quantity. Would this change current bahavior though?